### PR TITLE
types: return maxUint32 - 1 when overflow in convertToUint func

### LINKED ISSUE #29330: TiDB is inserting null when value >4294967295 (max value of unsigned int)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

```sql
drop table if exists  t1;
create table t1 (this int unsigned);
insert ignore into t1 values ('5000000000');
select * from t1
--- mysql inserting max value of unsigned int 4294967295
-- TiDB inserting null
```

### 2. What did you expect to see? (Required)
```sql
mysql> select * from t1;
+------------+
| this       |
+------------+
| 4294967295 |
+------------+
1 row in set (0.00 sec)
```
### 3. What did you see instead (Required)
```sql
mysql> select * from t1;
+------+
| this |
+------+
| NULL |
+------+
1 row in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```sql
+-------------------------+--------------------------------------------------------------------------+
| Variable_name           | Value                                                                    |
+-------------------------+--------------------------------------------------------------------------+
| innodb_version          | 5.6.25                                                                   |
| protocol_version        | 10                                                                       |
| tidb_analyze_version    | 2                                                                        |
| tidb_row_format_version | 2                                                                        |
| tls_version             | TLSv1,TLSv1.1,TLSv1.2                                                    |
| version                 | 5.7.25-TiDB-v5.2.1                                                       |
| version_comment         | TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible |
| version_compile_machine | x86_64                                                                   |
| version_compile_os      | osx10.8                                                                  |
+-------------------------+--------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #29330

Problem Summary: return maxUint32 - 1 when overflow in convertToUint func

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Defined2014** [comment]
/cc @kennytm @wshwsh12 @XuHuaiyu @morgo 
PTAL

**Review by morgo — COMMENTED** [review]
It LGTM, but sorry I don't understand types well. I will defer to other reviewers.

**Review by xhebox — APPROVED** [review]

**Comment by Defined2014** [comment]
/cc @wjhuang2016
PTAL

File: types/datum_test.go [hunk 1]
```diff
@@ -151,6 +151,30 @@ func TestToInt64(t *testing.T) {
...
 }
 
+func TestToUint32(t *testing.T) {
+	t.Parallel()
+
+	sc := new(stmtctx.StatementContext)
+	sc.IgnoreTruncate = true
+
+	// test overflow
+	var datum = NewUintDatum(5000000000)
```
Comment (wjhuang2016) [inline types/datum_test.go:161]: Please make this test more scalable and add more test cases for it.

Comment (Defined2014) [reply types/datum_test.go]: Done. PTAL @wjhuang2016

**Comment by xhebox** [comment]
@Defined2014 gentle ping on this

**Review by wjhuang2016 — APPROVED** [review]

**Comment by Defined2014** [comment]
@xhebox @wjhuang2016 Could you help me merge it?

**Pull request merged**